From fcfb70bfb2a9aae78e86a2505669068e7e511f79 Mon Sep 17 00:00:00 2001
From: Michael Brown <mcb30@ipxe.org>
Date: Mon, 23 Jan 2023 12:30:41 +0000
Subject: [PATCH] [arm] Inhibit linker warnings about an implied executable
 stack

Some versions of the 32-bit ARM linker seem to treat the absence of a
.note.GNU-stack section as implying an executable stack, and will
print a warning that this is deprecated behaviour.

Silence the warning by adding a .note.GNU-stack section to each
assembly file and retaining the sections in the Linux linker script.

Signed-off-by: Michael Brown <mcb30@ipxe.org>
---
 src/arch/arm32/core/setjmp.S     | 1 +
 src/arch/arm32/libgcc/lldivmod.S | 1 +
 src/arch/arm32/libgcc/llshift.S  | 1 +
 src/scripts/linux.lds            | 5 +----
 4 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/arch/arm32/core/setjmp.S b/src/arch/arm32/core/setjmp.S
index 7e7b0fe5..1e9e8202 100644
--- a/src/arch/arm32/core/setjmp.S
+++ b/src/arch/arm32/core/setjmp.S
@@ -1,5 +1,6 @@
 FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL )
 
+	.section ".note.GNU-stack", "", %progbits
 	.text
 	.arm
 
diff --git a/src/arch/arm32/libgcc/lldivmod.S b/src/arch/arm32/libgcc/lldivmod.S
index 910be4b7..746fa8fd 100644
--- a/src/arch/arm32/libgcc/lldivmod.S
+++ b/src/arch/arm32/libgcc/lldivmod.S
@@ -1,5 +1,6 @@
 FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL )
 
+	.section ".note.GNU-stack", "", %progbits
 	.text
 	.thumb
 
diff --git a/src/arch/arm32/libgcc/llshift.S b/src/arch/arm32/libgcc/llshift.S
index cc16e261..c1b51e77 100644
--- a/src/arch/arm32/libgcc/llshift.S
+++ b/src/arch/arm32/libgcc/llshift.S
@@ -1,5 +1,6 @@
 FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL )
 
+	.section ".note.GNU-stack", "", %progbits
 	.text
 	.arm
 
diff --git a/src/scripts/linux.lds b/src/scripts/linux.lds
index afc0128f..53e884ec 100644
--- a/src/scripts/linux.lds
+++ b/src/scripts/linux.lds
@@ -81,16 +81,13 @@ SECTIONS {
 	_assert = ASSERT ( ( _weak == _eweak ), ".weak is non-zero length" );
 
 	/*
-	 * Dispose of the comment and note sections to make the link map
-	 * easier to read
+	 * Dispose of unwanted sections to make the link map easier to read
 	 *
 	 */
 
 	/DISCARD/ : {
 		*(.comment)
 		*(.comment.*)
-		*(.note)
-		*(.note.*)
 		*(.rel)
 		*(.rel.*)
 		*(.discard)
-- 
2.39.5

